pandas/core/frame.py

```diff
@@ -2748,13 +2753,26 @@ def set_index(self, keys, drop=True, append=False, inplace=False,
         >>> indexed_df = df.set_index(['A', 'B'])
         >>> indexed_df2 = df.set_index(['A', [0, 1, 2, 0, 1, 2]])
         >>> indexed_df3 = df.set_index([[0, 1, 2, 0, 1, 2]])
+        >>> indexed_df4 = df.set_index(df.columns[:2])
 
         Returns
         -------
         dataframe : DataFrame
         """
         if not isinstance(keys, list):
-            keys = [keys]
+            if isinstance(keys, Index):
+                # if the index is a slice of the column index, treat it like
+                # a list of column labels; otherwise, treat it like a new index
+                keys_base = keys.base
```

Comment:
this is not a reliable approach, nor should you be using the private ``.base``.

I am not convinced we can do this in a reliable way.